File: LeetCodeRecord/ACFirst/92-ReverseBetween.java
package ACFirst;

class ReverseBetween {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
    
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy;
        curr.next = head;

        for (int i = 1; i < m; i++) {
            curr = curr.next;
        }

        ListNode prev = null;
        ListNode tail = curr.next;
        
        for (int i = 0; i <= n - m; i++) {
            ListNode temp = tail.next;
            tail.next = prev;
            prev = tail;
            tail = temp;
        }

        curr.next.next = tail;
        curr.next = prev;

        return dummy.next;
    }

    public ListNode reverseBetween2(ListNode head, int m, int n) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy;
        curr.next = head;

        for (int i = 1; i < m; i++) {
            curr = curr.next;
        }

        ListNode prev = curr;
        curr = curr.next;

        for (int i = 1; i <= n - m; i++) {
            ListNode temp1Node = curr.next.next;
            ListNode temp2Node = curr.next;
            curr.next.next = prev.next;
            prev.next = temp2Node;
            curr.next = temp1Node;
        }

        return dummy.next;
    }


    private ListNode tailNext = null;

    public ListNode reverseBegin(ListNode head, int len) {
        if (head == null || head.next == null) {
            return head;
        }

        if (len == 1) {
            tailNext = head.next;
            return head;
        }

        ListNode ret = this.reverseBegin(head.next, len - 1);
        head.next.next =  head;
        head.next = tailNext;

        return ret;

    }

    public ListNode reverseBetween3(ListNode head, int m, int n) {
        if (head == null || head.next == null) {
            return head;
        }

        if (m == n) {
            return head;
        }

        if (m == 1) {
            return this.reverseBegin(head, n);
        }

        ListNode ret = this.reverseBetween3(head.next, m - 1, n - 1);
        
        head.next = ret;

        return head;
    }
}

File: Top200/92-ReverseBetween.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	dummy := &ListNode{Val: -1}
	curr := dummy
	curr.Next = head

	for i := 1; i < left; i++ {
		curr = curr.Next
	}

	var prev *ListNode = nil
	tail := curr.Next

	for i := 0; i <= right-left; i++ {
		temp := tail.Next
		tail.Next = prev
		prev = tail
		tail = temp
	}

	curr.Next.Next = tail
	curr.Next = prev

	return dummy.Next
}

func reverseBetween2(head *ListNode, left int, right int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	dummy := &ListNode{Val: -1}
	curr := dummy
	curr.Next = head

	for i := 1; i < left; i++ {
		curr = curr.Next
	}

	prev := curr
	curr = curr.Next

	for i := 0; i < right-left; i++ {
		tempNode1 := curr.Next.Next
		tempNode2 := curr.Next
		curr.Next.Next = prev.Next
		prev.Next = tempNode2
		curr.Next = tempNode1
	}

	return dummy.Next
}
